File: linked-list/sort-list.java
// https://leetcode.com/problems/sort-list/

// Java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode slow = head, fast = head, prev = null;

        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        prev.next = null;
            
        ListNode l1 = sortList(head);
        ListNode l2 = sortList(slow);
            
        return merge(l1, l2);
    }


    public ListNode merge(ListNode list1, ListNode list2){
        ListNode head = new ListNode();
        ListNode res = head;

        while (list1 != null && list2 != null){

            if (list1.val < list2.val) {
                head.next = list1;
                list1 = list1.next;
            }

            else {
                head.next = list2;
                list2 = list2.next;
            }

            head = head.next;
        }

        head.next = (list1 != null) ? list1 : list2;

        return res.next;
    }
}

// the issue with sorting here is that it is not easy to access each of the nodes, and swap
// also we need to try to do in O(1) space, so it is trivial to store the values in pq n first traveral then re-edit the list in scond traversal
// use mergesort which is nlogn

File: linked_list/sort-list.js
// https://leetcode.com/problems/sort-list/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function (head) {
  let vals = [],
    dummy = head,
    res = head;

  while (head) {
    vals.push(head.val);
    head = head.next;
  }

  vals.sort((a, b) => a - b);

  for (let val of vals) {
    dummy.val = val;
    dummy = dummy.next;
  }

  return res;
};

// Time: O(nlogn) -> default sort is usually nlogn
// Space: O(n) -> create an array

// instead of sorting the list itself, i can sort the values in a normal array, and then rewrute the list with the values
// ill just iterate through the list and append the values to the list
// sort the list
// iterate the list again but now overwrite the values
